File: testing/Android/app/src/main/java/com/styleconverter/test/style/sizing/AspectRatioValue.kt
package com.styleconverter.test.style.sizing

// Phase 3 primitive: normalize every aspect-ratio IR shape into a single value
// type. Unlike Width/Height the aspect-ratio property has its own wire shape
// (not an IRLength) so it gets its own extractor rather than reusing the
// LengthValue pipeline.
//
// Reference fixtures (examples/properties/sizing/aspect-ratio.json):
//   {"ratio":{"w":16.0,"h":9.0},"normalizedRatio":1.777…}  — two-number form
//   {"ratio":{"value":1.5},"normalizedRatio":1.5}          — single-number form
//   "auto"                                                  — bare string
//   {"ratio":{"auto":true,"w":16.0,"h":9.0}, …}             — auto with fallback

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonPrimitive

/**
 * CSS aspect-ratio value. `ratio` is width/height (>0) when an explicit ratio
 * is present; when only `auto` is specified we use 0.0 as a sentinel and set
 * [isAuto]. The `auto <ratio>` form sets both [isAuto] and a positive [ratio]
 * — the platform may choose to honour either.
 */
data class AspectRatioValue(
    // Numeric width/height ratio. 0.0 when author wrote bare `auto` only.
    val ratio: Double,
    // True when the CSS `auto` keyword was present. Bare `auto` → ratio=0.0.
    val isAuto: Boolean = false,
)

/**
 * Normalize an IR aspect-ratio data element. Never throws. Returns null when
 * the shape is entirely unrecognised; that lets callers treat it the same as
 * "property not specified" rather than wire a distinct Unknown sentinel in.
 */
fun extractAspectRatio(data: JsonElement?): AspectRatioValue? {
    if (data == null) return null
    // Shape 3: bare string "auto" collapses the whole property.
    if (data is JsonPrimitive && data.isString && data.content == "auto") {
        // ratio=0.0 is our sentinel for "auto-only" — the Applier skips
        // aspectRatio() and leaves Compose to auto-size.
        return AspectRatioValue(ratio = 0.0, isAuto = true)
    }
    if (data !is JsonObject) return null
    // normalizedRatio is pre-computed by the parser for both the two-number
    // and single-number shapes. Prefer it when available so we're byte-for-byte
    // consistent with the generator's notion of the ratio.
    val normalized = data["normalizedRatio"]?.jsonPrimitive?.doubleOrNull
    // ratio sub-object carries the raw components. We read the `auto` flag
    // from here for the "auto <ratio>" variant (shape 4).
    val ratioObj = data["ratio"] as? JsonObject
    // IR emits `auto:true` as a real JSON boolean — accept booleans first,
    // then fall through to stringy "true"/numeric 1.0 for hand-written JSON.
    val autoFlag = ratioObj?.get("auto")?.jsonPrimitive?.let { prim ->
        prim.booleanOrNull
            ?: (if (prim.isString) prim.content == "true" else prim.doubleOrNull == 1.0)
    } ?: false
    if (normalized != null) {
        return AspectRatioValue(ratio = normalized, isAuto = autoFlag)
    }
    // Fallback path: compute ratio from w/h when normalizedRatio is missing.
    // Defensive — current IR always emits normalizedRatio, but hand-written
    // fixtures might not.
    if (ratioObj != null) {
        val w = ratioObj["w"]?.jsonPrimitive?.doubleOrNull
        val h = ratioObj["h"]?.jsonPrimitive?.doubleOrNull
        val single = ratioObj["value"]?.jsonPrimitive?.doubleOrNull
        if (single != null && single > 0.0) {
            return AspectRatioValue(ratio = single, isAuto = autoFlag)
        }
        if (w != null && h != null && h > 0.0) {
            return AspectRatioValue(ratio = w / h, isAuto = autoFlag)
        }
    }
    return null
}

File: runtimes/web/src/engine/sizing/AspectRatioValue.ts
// AspectRatioValue.ts — typed shape + extractor for the CSS `aspect-ratio`
// property.  The IR emits three distinct shapes (confirmed against the
// sizing/aspect-ratio.json fixture after running the Kotlin converter):
//   "auto"                                              -> { isAuto: true,  ratio: 0 }
//   { ratio:{w,h},          normalizedRatio: w/h }      -> { isAuto: false, ratio: w/h }
//   { ratio:{value:N},      normalizedRatio: N }        -> { isAuto: false, ratio: N }
//   { ratio:{auto,w,h},     normalizedRatio: w/h }      -> { isAuto: false, ratio: w/h }
// The last form corresponds to `aspect-ratio: auto 16/9` — CSS still applies
// the explicit ratio when height isn't otherwise constrained, so we treat it
// as a concrete ratio rather than auto.

// Output consumed by the SizeApplier to emit `aspectRatio: '<ratio>'` inline.
export interface AspectRatioValue {
  // 0 when isAuto is true (sentinel); width / height otherwise.
  ratio: number;
  // Set when the IR emitted the bare string "auto" (no explicit ratio).
  isAuto: boolean;
}

// Pure extractor — returns null for anything we don't recognise so the caller
// can simply skip emitting an aspectRatio style.  Never throws.
export function extractAspectRatio(data: unknown): AspectRatioValue | null {
  // Shape #1 — bare string "auto".
  if (data === 'auto') return { ratio: 0, isAuto: true };
  // Everything else must be an object envelope.
  if (typeof data !== 'object' || data === null) return null;
  // Cast once so subsequent field accesses stay readable.
  const d = data as Record<string, unknown>;
  // Preferred path — Kotlin already did the division for us, so trust it.
  if (typeof d.normalizedRatio === 'number') {
    return { ratio: d.normalizedRatio, isAuto: false };
  }
  // Fallback path — compute from the nested { w, h } pair if present.
  const r = d.ratio as Record<string, unknown> | undefined;
  if (r && typeof r === 'object') {
    // w/h pair takes precedence over the single-number `value` form.
    if (typeof r.w === 'number' && typeof r.h === 'number' && r.h !== 0) {
      return { ratio: (r.w as number) / (r.h as number), isAuto: false };
    }
    // Single numeric ratio (e.g. `aspect-ratio: 1.5`).
    if (typeof r.value === 'number') {
      return { ratio: r.value, isAuto: false };
    }
  }
  // Unrecognised shape — caller should not emit anything.
  return null;
}
